src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java

```diff
@@ -409,40 +413,48 @@ public Source(Commands commands,
-
-      // fake shortcuts for commands which we handle at a lower level
-      commands.goToHelp().setShortcut(new KeyboardShortcut("F1", KeyCodes.KEY_F1, KeyboardShortcut.NONE));
-      commands.goToDefinition().setShortcut(new KeyboardShortcut("F2", KeyCodes.KEY_F2, KeyboardShortcut.NONE));
-
-      // If tab has been disabled for auto complete by the user, set the "shortcut" to ctrl-space instead.
-      if (userPrefs_.tabCompletion().getValue() && !userPrefs_.tabKeyMoveFocus().getValue())
-         commands.codeCompletion().setShortcut(new KeyboardShortcut("Tab", KeyCodes.KEY_TAB, KeyboardShortcut.NONE));
-      else
-      {
-         KeySequence sequence = new KeySequence();
-         sequence.add(new KeyCombination("Ctrl+Space", KeyCodes.KEY_SPACE, KeyCodes.KEY_CTRL));
-         commands.codeCompletion().setShortcut(new KeyboardShortcut(sequence));
-      }
-      
-      events.addHandler(ShowContentEvent.TYPE, this);
-      events.addHandler(ShowDataEvent.TYPE, this);
-      events.addHandler(OpenObjectExplorerEvent.TYPE, this);
-
-      events.addHandler(CodeBrowserNavigationEvent.TYPE, this);
       
-      events.addHandler(CodeBrowserFinishedEvent.TYPE, this);
+      vimCommands_ = new SourceVimCommands();
 
-      events.addHandler(CodeBrowserHighlightEvent.TYPE, this);
+      view_.addTabClosingHandler(this);
+      view_.addTabCloseHandler(this);
+      view_.addTabClosedHandler(this);
+      view_.addTabReorderHandler(this);
+      view_.addSelectionHandler(this);
+      view_.addBeforeShowHandler(this);
 
-      events.addHandler(FileTypeChangedEvent.TYPE, new FileTypeChangedHandler()
+      events_.addHandler(EditPresentationSourceEvent.TYPE, this);
+      events_.addHandler(FileEditEvent.TYPE, this);
+      events_.addHandler(InsertSourceEvent.TYPE, this);
+      events_.addHandler(ShowContentEvent.TYPE, this);
+      events_.addHandler(ShowDataEvent.TYPE, this);
+      events_.addHandler(OpenObjectExplorerEvent.TYPE, this);
+      events_.addHandler(OpenPresentationSourceFileEvent.TYPE, this);
+      events_.addHandler(OpenSourceFileEvent.TYPE, this);
+      events_.addHandler(CodeBrowserNavigationEvent.TYPE, this);
+      events_.addHandler(CodeBrowserFinishedEvent.TYPE, this);
+      events_.addHandler(CodeBrowserHighlightEvent.TYPE, this);
+      events_.addHandler(SnippetsChangedEvent.TYPE, this);
+      events_.addHandler(NewDocumentWithCodeEvent.TYPE, this);
+
+      events_.addHandler(FileTypeChangedEvent.TYPE, new FileTypeChangedHandler()
       {
          public void onFileTypeChanged(FileTypeChangedEvent event)
          {
             manageCommands();
          }
       });
       
-      events.addHandler(SourceOnSaveChangedEvent.TYPE, 
+      events_.addHandler(NewDocumentWithCodeEvent.TYPE,
+                         new NewDocumentWithCodeEvent.Handler() {
+         @Override
+         public void onNewDocumentWithCode(NewDocumentWithCodeEvent event)
+         {
+            onNewDocumentWithCode(event);
```

Comment:
Doesn't this trigger an infinite recursive loop? (onNewDocumentWithCode calling itself?).